src/aspects/ec2-automated-shutdown/index.ts

```diff
@@ -0,0 +1,142 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { join } from 'path';
+import { Stack, Duration, IAspect } from 'aws-cdk-lib';
+import { Metric, Alarm, ComparisonOperator } from 'aws-cdk-lib/aws-cloudwatch';
+import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
+import { Instance, CfnInstance } from 'aws-cdk-lib/aws-ec2';
+import { PolicyStatement, Effect, AnyPrincipal } from 'aws-cdk-lib/aws-iam';
+import { IKey } from 'aws-cdk-lib/aws-kms';
+import { Code, Runtime } from 'aws-cdk-lib/aws-lambda';
+import { SnsEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
+import { Topic } from 'aws-cdk-lib/aws-sns';
+import { IConstruct } from 'constructs';
+import { SecureFunction } from '../../constructs/secure-function';
+import { LambdaConfiguration } from '../../interfaces/lambda-configuration';
+
+export interface EC2ShutdownProps {
+    /**
+     * The % CPU threshold below which the instance will be shut down
+     */
+    readonly cpuThreshold: number;
+
+    /**
+     * Optional Lambda configuration settings.
+     */
+    readonly lambdaConfiguration?: LambdaConfiguration;
+
+    /**
+     * Optional KMS Encryption Key to use for encrypting resources.
+     */
+    readonly encryption?: IKey;
+}
+
+/**
+ * Aspect that applies a mechanism to automatically shut down
+ * an EC2 instance when its CPU utilization falls below a specified threshold.
+ * Allows for cost optimization and the reduction of resources not being actively used.
+ */
+export class Ec2AutomatedShutdown implements IAspect {
+    private lambdaFunction!: SecureFunction;
+    private processedStacks: Set<Stack> = new Set();
+    private counter = 0;
+
+    constructor(private readonly props: EC2ShutdownProps) {
+        if (props.cpuThreshold <= 0 || props.cpuThreshold > 100) {
+            throw new Error('CPU threshold must be between 0 and 100');
+        }
+    }
+
+    visit(node: IConstruct): void {
+        if (!(node instanceof Instance) && !(node instanceof CfnInstance)) {
+            return;
+        }
+
+        const stack = Stack.of(node);
+
+        if (!this.processedStacks.has(stack)) {
+            this.ensureLambdaFunction(stack);
+            this.processedStacks.add(stack);
+        }
+
+        this.createShutdownMechanism(node);
+    }
+
+    private ensureLambdaFunction(stack: Stack): void {
+        if (!this.lambdaFunction) {
```

Comment:
I don't think this works the way we need. In this code we are assigning the Lambda Function to the Aspect class itself. So the first stack would get processed, and then set `this.lambdaFunction`. Then the second stack would then reference the `this.lambdaFunction` from the class. That would create a scenario where the Lambda Function from Stack1 would be used across future stacks (Stack2, 3, etc.). This would create an export/import for the Lambda function and is not something we want to do as it gets messy.

Let me know if I'm misunderstanding the code here but thats what it looks like from my perspective right now.

Check out the SecureSageMakerNotebook Aspect that got merged to see how I created a mapping between Stacks and ManagedPolicy.